In [18]:
import pandas as pd
from janitor import clean_names

In [2]:
pwd

'C:\\Users\\Teavera'

In [4]:
cd OneDrive\Документы\Codeclan\codeclan_homework_verashcherbina\week_12\day_3

C:\Users\Teavera\OneDrive\Документы\Codeclan\codeclan_homework_verashcherbina\week_12\day_3


In [5]:
pwd

'C:\\Users\\Teavera\\OneDrive\\Документы\\Codeclan\\codeclan_homework_verashcherbina\\week_12\\day_3'

# 1.
### Read in all 4 credit card transaction datasets and clean up the column names.

In [25]:
cardbase = pd.read_csv("data/CardBase.csv")
customerbase = pd.read_csv("data/CustomerBase.csv")
fraudbase = pd.read_csv("data/FraudBase.csv")
transactionbase = pd.read_csv("data/TransactionBase.csv")

In [33]:
cardbase = cardbase.clean_names()
cardbase

card_number card_family  credit_limit  cust_id
0    8638-5407-3631-8196     Premium        530000  CC67088
1    7106-4239-7093-1515        Gold         18000  CC12076
2    6492-5655-8241-3530     Premium        596000  CC97173
3    2868-5606-5152-5706        Gold         27000  CC55858
4    1438-6906-2509-8219    Platinum        142000  CC90518
..                   ...         ...           ...      ...
495  4708-4407-9601-6022     Premium        691000  CC64993
496  6042-2856-7280-2925        Gold         33000  CC26787
497  8706-3809-5167-3899     Premium        144000  CC32532
498  9540-8558-5897-5046     Premium        830000  CC90246
499  5587-7265-8118-7718        Gold         24000  CC37803

[500 rows x 4 columns]

In [34]:
customerbase = customerbase.clean_names()
customerbase

cust_id  age customer_segment customer_vintage_group
0     CC25034   35          Diamond                    VG1
1     CC59625   22          Diamond                    VG1
2     CC69314   33          Diamond                    VG1
3     CC67036   46          Diamond                    VG1
4     CC25597   44          Diamond                    VG1
...       ...  ...              ...                    ...
5669  CC53430   27         Platinum                    VG2
5670  CC31823   33         Platinum                    VG2
5671  CC15335   28         Platinum                    VG2
5672  CC81556   26         Platinum                    VG2
5673  CC38634   21         Platinum                    VG2

[5674 rows x 4 columns]

# 2. 
### Join the data containing card details and customer details by customer id, so that all records of card details and any matching records in customer details are kept.

In [35]:
cardcustomer = cardbase.merge(customerbase, how="left", on="cust_id")

In [36]:
cardcustomer

card_number card_family  credit_limit  cust_id  age  \
0    8638-5407-3631-8196     Premium        530000  CC67088   27   
1    7106-4239-7093-1515        Gold         18000  CC12076   48   
2    6492-5655-8241-3530     Premium        596000  CC97173   22   
3    2868-5606-5152-5706        Gold         27000  CC55858   30   
4    1438-6906-2509-8219    Platinum        142000  CC90518   50   
..                   ...         ...           ...      ...  ...   
495  4708-4407-9601-6022     Premium        691000  CC64993   34   
496  6042-2856-7280-2925        Gold         33000  CC26787   46   
497  8706-3809-5167-3899     Premium        144000  CC32532   28   
498  9540-8558-5897-5046     Premium        830000  CC90246   35   
499  5587-7265-8118-7718        Gold         24000  CC37803   21   

    customer_segment customer_vintage_group  
0            Diamond                    VG1  
1               Gold                    VG3  
2            Diamond                    VG1  
3            Diamond                    VG1  
4           Platinum                    VG2  
..               ...                    ...  
495          Diamond                    VG1  
496         Platinum                    VG2  
497         Platinum                    VG2  
498             Gold                    VG3  
499          Diamond                    VG1  

[500 rows x 7 columns]

# 3.
### Join the data containing fraud details with transaction details so all rows of both tables are kept. What does the resulting row number tell you?

In [39]:
fraudbase = fraudbase.clean_names()
fraudbase

transaction_id  fraud_flag
0     CTID50558449           1
1     CTID55936882           1
2     CTID63762180           1
3     CTID76723439           1
4     CTID21246201           1
..             ...         ...
104   CTID57993591           1
105   CTID91108283           1
106   CTID30494187           1
107   CTID51301522           1
108   CTID49517337           1

[109 rows x 2 columns]

In [40]:
transactionbase = transactionbase.clean_names()
transactionbase

transaction_id transaction_date       credit_card_id  transaction_value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     transaction_segment  
0                  SEG25  
1                  SEG16  
2                  SEG14  
3                  SEG17  
4                  SEG14  
...                  ...  
9995               SEG18  
9996               SEG12  
9997               SEG13  
9998               SEG14  
9999               SEG15  

[10000 rows x 5 columns]

In [72]:
fraudtransaction = transactionbase.merge(fraudbase, how="left", on="transaction_id")
fraudtransaction

transaction_id transaction_date       credit_card_id  transaction_value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     transaction_segment  fraud_flag  
0                  SEG25         NaN  
1                  SEG16         NaN  
2                  SEG14         NaN  
3                  SEG17         NaN  
4                  SEG14         NaN  
...                  ...         ...  
9995               SEG18         NaN  
9996               SEG12         NaN  
9997               SEG13         NaN  
9998               SEG14         NaN  
9999               SEG15         NaN  

[10000 rows x 6 columns]

# 4. ???
###  Perform a filtering join between cards and transactions so that only the card details for cards that have made at least one transaction are preserved.

In [121]:
semi_table = cardbase.merge(transactionbase, how = "inner", left_on='card_number', right_on='credit_card_id')
semi_table

card_number card_family  credit_limit  cust_id transaction_id  \
0     8638-5407-3631-8196     Premium        530000  CC67088   CTID85973577   
1     8638-5407-3631-8196     Premium        530000  CC67088   CTID13215166   
2     8638-5407-3631-8196     Premium        530000  CC67088   CTID41016745   
3     8638-5407-3631-8196     Premium        530000  CC67088   CTID82512584   
4     8638-5407-3631-8196     Premium        530000  CC67088   CTID97222954   
...                   ...         ...           ...      ...            ...   
9995  5587-7265-8118-7718        Gold         24000  CC37803   CTID76550379   
9996  5587-7265-8118-7718        Gold         24000  CC37803   CTID58221219   
9997  5587-7265-8118-7718        Gold         24000  CC37803   CTID73252195   
9998  5587-7265-8118-7718        Gold         24000  CC37803   CTID72271256   
9999  5587-7265-8118-7718        Gold         24000  CC37803   CTID81265917   

     transaction_date       credit_card_id  transaction_value  \
0           13-Feb-16  8638-5407-3631-8196              31592   
1           30-Oct-16  8638-5407-3631-8196               5894   
2            9-May-16  8638-5407-3631-8196              20925   
3           31-Aug-16  8638-5407-3631-8196              36556   
4           25-Nov-16  8638-5407-3631-8196              36715   
...               ...                  ...                ...   
9995        11-Oct-16  5587-7265-8118-7718              23298   
9996        12-Jan-16  5587-7265-8118-7718              49957   
9997         8-Dec-16  5587-7265-8118-7718              13508   
9998         5-Feb-16  5587-7265-8118-7718                624   
9999         5-Apr-16  5587-7265-8118-7718              14019   

     transaction_segment  
0                  SEG21  
1                  SEG12  
2                  SEG11  
3                  SEG11  
4                  SEG18  
...                  ...  
9995               SEG15  
9996               SEG18  
9997               SEG25  
9998               SEG24  
9999               SEG13  

[10000 rows x 9 columns]

In [122]:
in_both = cardbase['card_number'].isin(semi_table['credit_card_id'])
in_both

0      True
1      True
2      True
3      True
4      True
       ... 
495    True
496    True
497    True
498    True
499    True
Name: card_number, Length: 500, dtype: bool

In [123]:
result_cards = cardbase[in_both]
result_cards

card_number card_family  credit_limit  cust_id
0    8638-5407-3631-8196     Premium        530000  CC67088
1    7106-4239-7093-1515        Gold         18000  CC12076
2    6492-5655-8241-3530     Premium        596000  CC97173
3    2868-5606-5152-5706        Gold         27000  CC55858
4    1438-6906-2509-8219    Platinum        142000  CC90518
..                   ...         ...           ...      ...
495  4708-4407-9601-6022     Premium        691000  CC64993
496  6042-2856-7280-2925        Gold         33000  CC26787
497  8706-3809-5167-3899     Premium        144000  CC32532
498  9540-8558-5897-5046     Premium        830000  CC90246
499  5587-7265-8118-7718        Gold         24000  CC37803

[500 rows x 4 columns]

# 5.
### Read in hat_observations and split observation into two columns, hat_colour and hat_type.

In [110]:
hat_observations = pd.read_csv("data/hat_observations.csv")
hat_observations

day  month  year        observation  observation_count
0    25     10  2019        black,beret                  7
1    11      9  2019       blue,top hat                 10
2     4     12  2018    green,santa hat                  9
3    18      6  2018        white,beret                  9
4    26      9  2018    black,santa hat                  1
..  ...    ...   ...                ...                ...
95   15      8  2018  purple,bucket hat                  8
96   26      1  2018   green,bowler hat                  5
97   15     11  2018      black,top hat                  7
98   14      9  2019       green,helmet                  4
99    1     10  2018        blue,fedora                  4

[100 rows x 5 columns]

In [112]:
hat_observations[["hat_colour", "hat_type"]] = hat_observations["observation"].str.split(",", expand=True);
hat_observations = hat_observations.drop(columns="observation")
hat_observations

day  month  year  observation_count hat_colour    hat_type
0    25     10  2019                  7      black       beret
1    11      9  2019                 10       blue     top hat
2     4     12  2018                  9      green   santa hat
3    18      6  2018                  9      white       beret
4    26      9  2018                  1      black   santa hat
..  ...    ...   ...                ...        ...         ...
95   15      8  2018                  8     purple  bucket hat
96   26      1  2018                  5      green  bowler hat
97   15     11  2018                  7      black     top hat
98   14      9  2019                  4      green      helmet
99    1     10  2018                  4       blue      fedora

[100 rows x 6 columns]

# 6 ???
### Unite day, month, and year columns into a column called date using a suitable separator. Then find the date where the most berets were observed.

TypeError: unhashable type: 'list'

# Extension
## Question 1
### Can you join all 4 credit card datasets together so that you’re left with a dataset that looks like below with 109 rows and 12 columns?

In [77]:
test1 = transactionbase.merge(fraudbase, how="inner", on="transaction_id")
test1

transaction_id transaction_date       credit_card_id  transaction_value  \
0     CTID50558449         6-May-16  4246-1369-3659-8804               6984   
1     CTID55936882        29-Nov-16  1336-9200-1264-2551              34367   
2     CTID63762180         5-Dec-16  8528-6154-7390-5081              44550   
3     CTID76723439        15-Sep-16  7908-2695-7391-7499              48275   
4     CTID21246201        29-Feb-16  2524-4184-5908-6750              35751   
..             ...              ...                  ...                ...   
104   CTID57993591         2-Apr-16  6697-9358-9213-4896              31486   
105   CTID91108283        17-Jul-16  9207-1270-6690-4905              36706   
106   CTID30494187        31-Dec-16  9018-5320-5729-5393              14586   
107   CTID51301522        28-Jan-16  3620-5235-2101-3391              22354   
108   CTID49517337        30-Sep-16  9604-6821-2861-8084              35688   

    transaction_segment  fraud_flag  
0                 SEG23           1  
1                 SEG21           1  
2                 SEG15           1  
3                 SEG16           1  
4                 SEG25           1  
..                  ...         ...  
104               SEG24           1  
105               SEG11           1  
106               SEG20           1  
107               SEG23           1  
108               SEG24           1  

[109 rows x 6 columns]

In [78]:
result = cardcustomer.merge(test1, how="inner", left_on='card_number', right_on='credit_card_id').drop(columns='credit_card_id')
result

card_number card_family  credit_limit  cust_id  age  \
0    7214-4915-6387-5429    Platinum        172000  CC72302   22   
1    6239-8641-8524-9441        Gold         26000  CC18438   41   
2    6228-9033-3508-8121        Gold         33000  CC36771   50   
3    6228-9033-3508-8121        Gold         33000  CC36771   50   
4    1466-1600-9312-7604     Premium        249000  CC60926   24   
..                   ...         ...           ...      ...  ...   
104  1947-8327-3848-6581    Platinum        113000  CC11690   24   
105  9018-5320-5729-5393     Premium        444000  CC54582   29   
106  2949-1972-1300-6680        Gold         45000  CC68567   31   
107  5384-8278-5616-5502     Premium        361000  CC83124   35   
108  9542-6082-9630-9970    Platinum        159000  CC66351   22   

    customer_segment customer_vintage_group transaction_id transaction_date  \
0            Diamond                    VG1   CTID87470159        11-Mar-16   
1            Diamond                    VG1   CTID40881434        10-Sep-16   
2            Diamond                    VG1   CTID42980186        17-Feb-16   
3            Diamond                    VG1   CTID33699337        16-Apr-16   
4            Diamond                    VG1   CTID25962688        18-Aug-16   
..               ...                    ...            ...              ...   
104          Diamond                    VG1   CTID16281374         6-Sep-16   
105             Gold                    VG3   CTID30494187        31-Dec-16   
106         Platinum                    VG2   CTID68584964        31-Dec-16   
107             Gold                    VG3   CTID38931999        28-Sep-16   
108         Platinum                    VG2   CTID88342446        21-Apr-16   

     transaction_value transaction_segment  fraud_flag  
0                32198               SEG21           1  
1                43482               SEG15           1  
2                45625               SEG22           1  
3                14442               SEG16           1  
4                45807               SEG20           1  
..                 ...                 ...         ...  
104              34290               SEG21           1  
105              14586               SEG20           1  
106              18851               SEG22           1  
107              29079               SEG23           1  
108              32930               SEG14           1  

[109 rows x 12 columns]

### Read in exam_scores and transform it into long format with two new columns exam_question and score. Remove superfluous information from the values in exam_question.

In [125]:
exam_scores = pd.read_csv("data/exam_scores.csv")
exam_scores

id  exam_Q1  exam_Q2  exam_Q3  exam_Q4  exam_Q5  exam_Q6  exam_Q7  \
0      1       69       59       95       56       54       78       86   
1      2       37       93       56       75       20       55       20   
2      3       38       23       47       37       92       26       73   
3      4       49       79       80       58       23       45       21   
4      5       40       92       59       47       30       66       53   
..   ...      ...      ...      ...      ...      ...      ...      ...   
195  196       75       71       98       87       30       46       91   
196  197       66       60       79       98       37       48       85   
197  198       99       66       67       22       68       35       65   
198  199       22       99       53       66       39       85       59   
199  200       65       86       30       21       85       74       83   

     exam_Q8  exam_Q9  exam_Q10  
0         99       34        47  
1         93       75        70  
2         34       87        22  
3         51       25        56  
4         64       86        64  
..       ...      ...       ...  
195       40       43        54  
196       59       30        28  
197       21       60        28  
198       97       67        67  
199      100       90        33  

[200 rows x 11 columns]

In [129]:
exam_scores_melted = exam_scores.melt(
    id_vars=['id'], 
    var_name='exam_questions',
    value_name='exam_score'
)
exam_scores_melted

id exam_questions  exam_score
0       1        exam_Q1          69
1       2        exam_Q1          37
2       3        exam_Q1          38
3       4        exam_Q1          49
4       5        exam_Q1          40
...   ...            ...         ...
1995  196       exam_Q10          54
1996  197       exam_Q10          28
1997  198       exam_Q10          28
1998  199       exam_Q10          67
1999  200       exam_Q10          33

[2000 rows x 3 columns]